### DISCUSSION TITLE
Get a Random Subset

### DISCUSSION THREAD

**Discussion by SteveSande** [feathersjs/feathers#3335] [Q&A] [answered]
I am generating a REST API and I would like to have the ability to get a random subset. Similar functionality can be found at this [free dog API](https://dog.ceo/dog-api/documentation/random). What would be your recommendation on how to implement this in Feathers?

I thought about having something in the query string to trigger a before hook that would:
1. Get all the relevant id's.
2. Generate random numbers to correspond to indexes in an array of the id's.
3. Reconstruct the query to get documents for those id's.

While implementing that solution I got stuck trying to add another query property in querySyntax(). I couldn't seem to figure out the exact formatting to allow my query property.

Edit: To be clear [this is the documentation example](https://feathersjs.com/api/schema/typebox.html#querysyntax) that I think is an analog for adding a query property. In the example case an '$ilike' property is added.

**Comment by ericuldall** [comment]
Which database adapter are you using?

**Reply by SteveSande** [reply]
I am using MongoDB.

**Reply by ericuldall** [reply, marked as answer]
You can use a mongo aggregation pipeline to get your sample:
https://www.mongodb.com/docs/manual/reference/operator/aggregation/sample/

**Reply by SteveSande** [reply]
Oh thank you for pointing this out! I'm new to MongoDB. Hopefully I can try it out and get it working on the weekend.

**Reply by ericuldall** [reply]
Happy to help! If you run into any trouble feel free to @ me in a new discussion.

**Comment by noor-tg** [comment]
The simplest algorithm I thought of is to get a limited number of data by $limit and sort the general collection by random field with $sort

For example 
```ts
const field = random(['name', 'bio', 'age', 'created_at'])
const query {
    $limit: 30,
    $sort: {
       [field]: 1
    }
}

const result = client.service("dogs").find({query})
```
The `random` function can be made custom or from faker or any other way

**Reply by SteveSande** [reply]
This is an interesting solution. I could also generate a random value for $skip to increase randomness. Thanks for your thoughts.

In order to implement this on the server I would still need some sort of query property to act as a trigger. There is some code in the [documentation here](https://feathersjs.com/api/schema/typebox.html#querysyntax) that I tried to use to help add a 'random' query property, but I couldn't get it to work. I suppose if I can't get it to work then I can add it to my object schema. Or I can keep things simple and leave it up to the client to generate code for getting a random subset.
